### PR TITLE
Dockerization of exchange-sandbox

### PR DESCRIPTION
**Pull request by stevenewald**

### CODE REVIEW

File: Taskfile.yml [hunk 1]
```diff
@@ -69,3 +69,33 @@ tasks:
       - task: init-linter
       - task: init-analyzer
       - task: init-web
+
+  exchange-docker-build:
+    cmds:
+      - sudo docker build -t nutc-exchange:latest . -f exchange/Dockerfile
```
Comment (egelja) [inline Taskfile.yml:75]: No sudo

File: exchange/CMakePresets.json [hunk 2]
```diff
@@ -65,7 +65,7 @@
       "name": "flags-linux",
       "hidden": true,
       "cacheVariables": {
-        "CMAKE_CXX_FLAGS": "-D_FORTIFY_SOURCE=3 -fstack-protector-strong -fcf-protection=full -fstack-clash-protection -Wall -Wextra -Wpedantic -Wconversion -Wsign-conversion -Wcast-qual -Wformat=2 -Wundef -Werror=float-equal -Wshadow -Wcast-align -Wunused -Wnull-dereference -Wdouble-promotion -Wimplicit-fallthrough -Wextra-semi -Woverloaded-virtual -Wnon-virtual-dtor -Wold-style-cast",
```
Comment (egelja) [inline exchange/CMakePresets.json:68]: This shouldn't be necessary

File: exchange/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+COPY ./exchange /app/exchange
+RUN cmake --preset=ci-docker \
+    && cmake --build build --config Release -j
+
+# Main stage for exchange
+FROM python:3.11-slim as run-exchange
+
+RUN apt update
+RUN apt install -y rabbitmq-server
+RUN pip install numpy pandas polars scipy scikit-learn
```
Comment (egelja) [inline exchange/Dockerfile:48]: Each RUN command is a new layer

File: wrapper/CMakeLists.txt [hunk 4]
```diff
@@ -30,11 +30,13 @@ find_package(Python COMPONENTS Interpreter Development REQUIRED)
 find_package(pybind11 REQUIRED)
 
 # Git version tracking
-FetchContent_Declare(cmake_git_version_tracking
-  GIT_REPOSITORY https://github.com/andrew-hardin/cmake-git-version-tracking.git
-  GIT_TAG 9b5fc5088b4089ff2adc20d607976b9923e3d737
-)
-FetchContent_MakeAvailable(cmake_git_version_tracking)
+if(USE_GIT_VERSION_TRACKING)
```
Comment (egelja) [inline wrapper/CMakeLists.txt:33]: Why

File: wrapper/CMakePresets.json [hunk 5]
```diff
@@ -65,7 +65,7 @@
       "name": "flags-linux",
       "hidden": true,
       "cacheVariables": {
-        "CMAKE_CXX_FLAGS": "-D_FORTIFY_SOURCE=3 -fstack-protector-strong -fcf-protection=full -fstack-clash-protection -Wall -Wextra -Wpedantic -Wconversion -Wsign-conversion -Wcast-qual -Wformat=2 -Wundef -Werror=float-equal -Wshadow -Wcast-align -Wunused -Wnull-dereference -Wdouble-promotion -Wimplicit-fallthrough -Wextra-semi -Woverloaded-virtual -Wnon-virtual-dtor -Wold-style-cast",
```
Comment (egelja) [inline wrapper/CMakePresets.json:68]: Should not be necessary

File: wrapper/src/main.cpp [hunk 6]
```diff
@@ -14,6 +13,10 @@
 #include <string>
 #include <tuple>
 
+#ifndef NO_GIT_VERSION_TRACKING
```
Comment (egelja) [inline wrapper/src/main.cpp:16]: Why

**Review by egelja — CHANGES_REQUESTED** [review]
Looks very good

File: wrapper/CMakePresets.json (hunk 5 above)
Comment (stevenewald) [reply wrapper/CMakePresets.json:68]: It is to build on docker on macos. The flag is not worth the annoyance to me+gavin+max imo

File: Taskfile.yml (hunk 1 above)
Comment (egelja) [reply Taskfile.yml:75]: https://stackoverflow.com/a/73258705

Comment (stevenewald) [reply Taskfile.yml:75]: Does not work

File: exchange/src/process_spawning/spawning.cpp [hunk 7]
```diff
@@ -46,27 +46,37 @@ get_all_users()
     return res.get<glz::json_t::object_t>();
 }
 
+std::string
+quote_id(const std::string& id)
+{
+    std::string quote_id = std::string(id);
+    std::replace(quote_id.begin(), quote_id.end(), '-', ' ');
+    return quote_id;
+}
```
Comment (egelja) [inline exchange/src/process_spawning/spawning.cpp:55]: Just take a copy and the user can `std::move` if possible

```suggestion
quote_id(std::string id)
{
    std::replace(id.begin(), id.end(), '-', ' ');
    return id;
}
```

File: wrapper/src/main.cpp [hunk 8]
```diff
@@ -14,7 +17,7 @@
 #include <string>
 #include <tuple>
 
-static std::tuple<uint8_t, std::string, bool>
+static std::tuple<uint8_t, std::string, std::string, bool>
```
Comment (egelja) [inline wrapper/src/main.cpp:20]: make this a struct at this point

**Review by egelja — CHANGES_REQUESTED** [review]
almost done

File: Taskfile.yml (hunk 1 above)
Comment (egelja) [reply Taskfile.yml:75]: discussed in slack

**Comment by egelja** [comment]
steev

**Comment by stevenewald** [comment]
im busy ill get to it

**Comment by egelja** [comment]
On 11/15/2023 11:09, Steve wrote:
> im busy ill get to it

ok steev no worries sorry to bother you

-- 
Nikola Maruszewski
Don't Panic - The Hitchhiker's Guide to the Galaxy.
Northwestern University | Class of 2025
BS in Computer Science | MS in Computer Engineering
***@***.***
***@***.***
https://nikola.cx

File: exchange/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    && apt install -y rabbitmq-server \
+    && pip install numpy pandas polars scipy scikit-learn
+
+COPY --from=build-wrapper /app/wrapper/build/NUTC-client /bin/NUTC-client
+COPY --from=build-exchange /app/exchange/build/NUTC24 /bin/NUTC-exchange
+COPY ./exchange/scripts/docker-start.sh /app/docker-start.sh
+
+RUN chmod +x /bin/NUTC-client
+RUN chmod +x /bin/NUTC-exchange
+RUN chmod +X /app/docker-start.sh
```
Comment (egelja) [inline exchange/Dockerfile:56]:
```suggestion
RUN chmod +x /bin/NUTC-client /bin/NUTC-exchange /app/docker-start.sh
```

**Review by egelja — CHANGES_REQUESTED** [review]

**Review by egelja — APPROVED** [review]
lgtm

**Pull request merged**